### PR TITLE
Support for additional date/time input types in <InputDate> and support for DateOnly/TimeOnly bindings.

### PR DESCRIPTION
**Pull request by MackinnonBuck**
### Support for additional date/time input types in `<InputDate>` and support for `DateOnly`/`TimeOnly` bindings.
These changes add support for `month`, `datetime-local`, and `time` input types for the `<InputDate>` component. Binding support for `DateOnly` and `TimeOnly` has also been added.

**PR Description**
Before these changes, `<input type="datetime-local" ...>` could not bind to a `DateTime` without specifying a `@bind-format`. This is no longer a requirement for use.

Developers can also bind to the value of date/time inputs using the new `DateOnly` and `TimeOnly` types. Currently, this works by first parsing as a `DateTime` using the appropriate format, then converting to a `DateOnly` or `TimeOnly`. This allows someone to bind a `DateOnly` when using `datetime-local`, for instance, if they wish to ignore time information. *If this is not a scenario we want to support, we could change this behavior.*

One can also now specify a `type` attribute for `<InputDate>` if they wish to use a date/time input type other than "date". This keeps the API surface small, but we could alternatively introduce an `InputDateType` enum parameter to `<InputDate>` if that makes the feature more visible to developers.

**Validation**
- [x] Unit tests for new bind parsers/formatters.
- [x] E2E tests for `DateOnly`/`TimeOnly` bindings for date/datetime-local/month/time inputs.
- [x] E2E tests for `<InputDate>` improvements

Addresses #12376

### CODE REVIEW

File: src/Components/Components/src/BindConverter.cs [hunk 1]
```diff
@@ -488,6 +488,182 @@ private static string FormatDateTimeOffsetValueCore(DateTimeOffset value, Cultur
...
+        /// <returns>The formatted value.</returns>
+        [SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "Required to maintain compatibility")]
+        public static string FormatValue(DateOnly value, string format, CultureInfo? culture = null) => FormatDateOnlyValueCore(value, format, culture);
+
+        private static string FormatDateOnlyValueCore(DateOnly value, string? format, CultureInfo? culture)
+        {
+            if (format != null)
+            {
+                // We convert to a DateTime so formatting doesn't throw if the format includes time information
+                return value.ToDateTime(TimeOnly.MinValue).ToString(format, culture ?? CultureInfo.CurrentCulture);
```
Comment (MackinnonBuck) [inline src/Components/Components/src/BindConverter.cs:519]: We don't have to do this if we decide not to support binding a value of type `DateOnly` to inputs with time information (e.g. `datetime-local`). Same question applies for `TimeOnly`.

File: src/Components/Web.JS/src/Rendering/Events/EventTypes.ts [hunk 2]
```diff
@@ -258,9 +258,10 @@ function normalizeTimeBasedValue(element: HTMLInputElement): string {
   const type = element.type;
   switch (type) {
     case 'date':
-    case 'datetime-local':
     case 'month':
       return value;
+    case 'datetime-local':
+      return value.length === 16 ? value + ':00' : value; // Convert yyyy-MM-ddTHH:mm to yyyy-MM-ddTHH:mm:00
```
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Rendering/Events/EventTypes.ts:264]: If a `step` is not specified, seconds are not included in the format, so we need to add them here for conversion to succeed. This was the cause of the problem illustrated in the code sample in #12376.

File: src/Components/Web/src/Forms/InputDate.cs [hunk 3]
```diff
@@ -45,17 +69,14 @@ protected override void BuildRenderTree(RenderTreeBuilder builder)
...
+            {
+                DateTime dateTimeValue => BindConverter.FormatValue(dateTimeValue, _format, CultureInfo.InvariantCulture),
+                DateTimeOffset dateTimeOffsetValue => BindConverter.FormatValue(dateTimeOffsetValue, _format, CultureInfo.InvariantCulture),
+                DateOnly dateOnlyValue => BindConverter.FormatValue(dateOnlyValue, _format, CultureInfo.InvariantCulture),
+                TimeOnly timeOnlyValue => BindConverter.FormatValue(timeOnlyValue, _format, CultureInfo.InvariantCulture),
+                _ => string.Empty, // Handles null for Nullable<DateTime>, etc.
+            };
 
         /// <inheritdoc />
         protected override bool TryParseValueFromString(string? value, [MaybeNullWhen(false)] out TValue result, [NotNullWhen(false)] out string? validationErrorMessage)
```
Comment (MackinnonBuck) [inline src/Components/Web/src/Forms/InputDate.cs:82]: This might not be directly related to this PR, but is there a reason we are manually checking `TValue` and calling the appropriate converter rather than relying on `BindConverter.TryConvertTo<T>()`? The only reason I can think of would be so that we can throw if `TValue` is not a date/time type. But if that's the case, couldn't we check that in the constructor and throw right away rather than waiting until `TryParseValueFromString` and performing the check every time? Relying on `TryConvertTo<T>()` would also allow us to remove ~80 LOC from this file. If there was another reason we're doing things this way, I'm completely fine with leaving it as is.

@SteveSandersonMS @pranavkm

**Comment by MackinnonBuck** [comment]
Something else to consider is whether we want to change the default parsing error message when the `type` attribute in `<InputDate>` is something other than "date".

https://github.com/dotnet/aspnetcore/pull/34594/files#diff-385121ec218042f5cd83d8b8f85c30fc76f8f2551473591b52295bffe03add9aR28

File: src/Components/Web/src/Forms/InputDate.cs [hunk 4]
```diff
@@ -30,12 +35,31 @@ public class InputDate<TValue> : InputBase<TValue>
         /// </summary>
         [DisallowNull] public ElementReference? Element { get; protected set; }
 
+        /// <inheritdoc />
+        public override Task SetParametersAsync(ParameterView parameters)
+        {
+            _typeAttributeValue = parameters.TryGetValue<string>("type", out var typeAttributeValue)
+                ? typeAttributeValue
+                : "date";
```
Comment (campersau) [inline src/Components/Web/src/Forms/InputDate.cs:43]: What do you think about using `"time"` by default when `typeof(TValue) == typeof(TimeOnly)`?

File: src/Components/Web/src/Forms/InputDate.cs [hunk 5]
```diff
@@ -30,12 +35,31 @@ public class InputDate<TValue> : InputBase<TValue>
...
+            _format = _typeAttributeValue switch
+            {
+                "date" => DateFormat,
+                "datetime-local" => DateTimeLocalFormat,
+                "month" => MonthFormat,
+                "time" => TimeFormat,
+                _ => throw new InvalidOperationException($"Unsupported 'type' attribute value '{_typeAttributeValue}'.")
+            };
+
+            return base.SetParametersAsync(parameters);
```
Comment (SteveSandersonMS) [inline src/Components/Web/src/Forms/InputDate.cs:54]: Rather than run `parameters.TryGetValue` to find `"type"` above, would it work to call `base.SetParametersAsync` first, and then read the `"type"` entry out of `AdditionalAttributes`?

File: src/Components/Web/src/Forms/InputDate.cs (hunk 4 above)
Comment (SteveSandersonMS) [reply src/Components/Web/src/Forms/InputDate.cs:43]: The design here is based on the developer specifying their chosen `type` attribute, or defaulting to `"date"`. In most cases it's orthogonal to the type of the .NET property being bound, because we convert automatically between all the different date/time .NET types (even if that involves discarding some information).

I would rather keep the rules simple, as they are in this PR. That is, the developer picks a `type` to control the UI appearance, and the associated bound .NET type is independent of that.

File: src/Components/Web/src/Forms/InputDate.cs (hunk 5 above)
Comment (SteveSandersonMS) [reply src/Components/Web/src/Forms/InputDate.cs:54]: Also, more broadly, should we consider *not* using the `type` attribute for this, but instead having a parameter like:

```cs
[Parameter] public InputDateType Type { get; set; } = InputDateType.Date;

enum InputDateType { Date, Time, DateTime, Month }
```

I'm asking because it's unclear that there's an advantage to coupling this to the HTML `type` strings. They won't appear in intellisense, and TBH their naming is nonobvious and hard to remember (specifically, `datetime-local` - local to the server or the client? It's not clear for something like Blazor Server).

Since it's never been possible to pass a regular HTML `type` attribute to this component, because we always overrode it, it wouldn't be breaking if we continued to ignore custom `type` attributes and used a separate .NET parameter instead.

Comment (SteveSandersonMS) [reply src/Components/Web/src/Forms/InputDate.cs:54]: I'm not 100% set either way on this and can see arguments in both directions. If you have a clear preference @MackinnonBuck then let's go with that since you've been thinking about it the most.

File: src/Components/test/E2ETest/Tests/FormsTest.cs [hunk 6]
```diff
@@ -249,6 +249,87 @@ public void InputDateInteractsWithEditContext_NullableDateTimeOffset()
...
+
+            // Validates on edit
+            Browser.Equal("valid", () => departureTimeInput.GetAttribute("class"));
+            departureTimeInput.SendKeys("06:43\t");
+            Browser.Equal("modified valid", () => departureTimeInput.GetAttribute("class"));
+
+            // Can become invalid
+            ApplyInvalidInputDateValue(".departure-time input", "01:234:56");
+            Browser.Equal("modified invalid", () => departureTimeInput.GetAttribute("class"));
+            Browser.Equal(new[] { "The DepartureTime field must be a date." }, messagesAccessor);
```
Comment (SteveSandersonMS) [inline src/Components/test/E2ETest/Tests/FormsTest.cs:267]: Should the default message be sensitive to the type? Should this say `must be a time`, for example?

File: src/Components/Web/src/Forms/InputDate.cs (hunk 3 above)
Comment (SteveSandersonMS) [reply src/Components/Web/src/Forms/InputDate.cs:82]: I'm not totally sure, but the first difference I noticed is that the logic here defines its own format string for the parsing, whereas `BindConverter.TryConvertTo<T>` doesn't let you specify any format string. Internally, `BindConverter.TryConvertTo<T>` passes `null` as the format to things like `DateTime.TryParse`.

Are you saying that the formats used by the browser happen to be parseable by `DateTime.TryParse` (and similar for `TimeOnly.TryParse`, etc.) using a `null` format in all cases, and therefore we don't need to pass a format string for parsing?

If that's the case, and everything still works if changing to `BindConverter.TryConvertTo<T>`, then that sounds great and I'd definitely support that. Doing the type compatibility check once up front also makes sense.

**Comment by SteveSandersonMS** [comment]
This is looking great! Superb job with all the tests too.

File: src/Components/Components/src/BindConverter.cs (hunk 1 above)
Comment (pranavkm) [reply src/Components/Components/src/BindConverter.cs:519]: It seems like we'd be hiding a possibly misconfigured app if we had this. What does the error experience look like if we did not do this? Does it throw a meaningful exception?

File: src/Components/Web.JS/src/Rendering/Events/EventTypes.ts (hunk 2 above)
Comment (pranavkm) [reply src/Components/Web.JS/src/Rendering/Events/EventTypes.ts:264]: Do you mind expanding the comment to include this?

Comment (pranavkm) [reply src/Components/Web.JS/src/Rendering/Events/EventTypes.ts:264]: Also, is there a better way to detect this or handle this rather than relying on the length of the value? It seems somewhat brittle.

File: src/Components/Web/src/Forms/InputDate.cs (hunk 5 above)
Comment (pranavkm) [reply src/Components/Web/src/Forms/InputDate.cs:54]: How do we feel about having a `[Parameter, EditorRequired] public string Type { get; set; }` where the `string` is one of the regular values? I'm somewhat ambivalent about adding a new enum type for this, but we should flag that specifying `Type` is required.

File: src/Components/Web.JS/src/Rendering/Events/EventTypes.ts (hunk 2 above)
Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Rendering/Events/EventTypes.ts:264]: If there is some web standards spec somewhere saying that the only possible formats are `yyyy-MM-ddTHH:mm` and `yyyy-MM-ddTHH:mm:ss` then the length check seems fine. But if things are more ambiguous, then yeah I agree - I guess we could do a regex looking for `^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}$` or whatever the correct syntax is.

File: src/Components/Web/src/Forms/InputDate.cs (hunk 5 above)
Comment (SteveSandersonMS) [reply src/Components/Web/src/Forms/InputDate.cs:54]: Why would specifying `Type` be required? For back-compat, we should default to `date`.

File: src/Components/test/E2ETest/Tests/FormsTest.cs (hunk 6 above)
Comment (SteveSandersonMS) [reply src/Components/test/E2ETest/Tests/FormsTest.cs:267]: Oh I see you commented about that elsewhere too.

File: src/Components/Components/src/BindConverter.cs (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/BindConverter.cs:519]: I kind of like the flexibility that you can use any UI style with any underlying data type, and we always do the best available conversion (including discarding information when applicable). If there were cases where this would actively surprise developers or give them incorrect results then we should reconsider, but it seems like we're always giving the developer the most information we can based on the data type they've chosen.

Or am I misunderstanding? Are there any cases where people might get undesirable results?

Comment (SteveSandersonMS) [reply src/Components/Components/src/BindConverter.cs:519]: One thing it would be great to check we've thought through is timezones. If you bind a `type="month"` to a `DateTime`, does it show up as a local datetime or a UTC one? Does the browser supply a string that represents a specific timezone (e.g., with the `Z` suffix for UTC) or does it just ambiguously say something like `2021-08` and if so what kind of `DateTime` do we supply?

I would *hope* that the `DateTime` would represent midnight at the start of the month in UTC, and equivalent for picking a day. For `TimeOnly`, it looks like there is no timezone so that's probably correct, but again it would be good to check that even if you bind a `type="datetime-local"` to a `TimeOnly` that the hour/minute numbers entered by the user are what show up as the hour/minute numbers on the `TimeOnly` regardless of the browser and server timezones.

If there's anything difficult or risky about any of these cases, then I'd definitely be open to just not supporting them.

Comment (MackinnonBuck) [reply src/Components/Components/src/BindConverter.cs:519]: The exception that gets thrown is from `DateOnly.ToString()`:

```
Unhandled exception rendering component: Input string was not in a correct format.
      System.FormatException: Input string was not in a correct format.
         at System.DateTimeFormat.IsValidCustomDateFormat(ReadOnlySpan`1 format, Boolean throwOnError) in System.Private.CoreLib.dll:token 0x600106e+0x124
         at System.DateOnly.ToString(String format, IFormatProvider provider) in System.Private.CoreLib.dll:token 0x6000dda+0xf1
         at Microsoft.AspNetCore.Components.BindConverter.FormatDateOnlyValueCore(DateOnly value, String format, CultureInfo culture) in C:\aspnetcore\src\Components\Components\src\BindConverter.cs:line 520
...
```

Not ideal perhaps but it would probably get the idea across about what the problem was. I'm good with changing this functionality if we know there isn't a scenario where you would want to bind a `DateOnly` to a `type=datetime-local` input, etc.

File: src/Components/Web.JS/src/Rendering/Events/EventTypes.ts (hunk 2 above)
Comment (MackinnonBuck) [reply src/Components/Web.JS/src/Rendering/Events/EventTypes.ts:264]: Sure, I could change the logic to see if the `step` property is defined and use that to determine whether we need to normalize the format like this. The case for `time` inputs works the same way so I can change that one too.

File: src/Components/Web/src/Forms/InputDate.cs (hunk 3 above)
Comment (MackinnonBuck) [reply src/Components/Web/src/Forms/InputDate.cs:82]: It does look like the browser format is parseable by `DateTime`, etc. using a `null` format in all cases. Everything still works after changing it to use `BindConverter.TryConvertTo<T>()`. But I'm happy again to leave it as-is if we just don't want to risk breaking some weird edge case or something (although I don't foresee that happening especially if we throw in the constructor when `TValue` is something weird).

File: src/Components/Components/src/BindConverter.cs (hunk 1 above)
Comment (pranavkm) [reply src/Components/Components/src/BindConverter.cs:519]: There's a couple places in the framework that catch FormatException and wrap it in a more specific error. I'd be open to that.

File: src/Components/test/E2ETest/Tests/FormsTest.cs (hunk 6 above)
Comment (MackinnonBuck) [reply src/Components/test/E2ETest/Tests/FormsTest.cs:267]: That was a question I had as well. I think it probably should be sensitive to the input type. I'll move forward with this change for now.

File: src/Components/Web/src/Forms/InputDate.cs (hunk 3 above)
Comment (SteveSandersonMS) [reply src/Components/Web/src/Forms/InputDate.cs:82]: In that case, simplifying this sounds great. The test coverage looks good so I'm not aware there will be other edge cases. If you can specifically anticipate any then it's worth discussing, but otherwise it seems fine.

File: src/Components/Components/src/BindConverter.cs (hunk 1 above)
Comment (MackinnonBuck) [reply src/Components/Components/src/BindConverter.cs:519]: I didn't mean to ignore your comments, @SteveSandersonMS - I had to refresh the page for them to show up 😅 

The browser just supplies a "yyyy-MM" string for `type="month"` inputs, and it does appear that the `DateTime` we get is the start of the month at midnight. Same also applies for picking a day for `type="date"`. Since `DateTime` doesn't have time zones (I believe that's what `DateTimeOffset` would be used for), I don't think we'll have time zone issues here.

I personally don't see it as being a problem to support these cases. But, if they truly only occur when the developer has made a mistake by specifying the wrong input type or binding to the wrong date/time type, maybe it would be better to throw an error 🤷‍♂️

File: src/Components/Web.JS/src/Rendering/Events/EventTypes.ts (hunk 2 above)
Comment (MackinnonBuck) [reply src/Components/Web.JS/src/Rendering/Events/EventTypes.ts:264]: Once again, didn't mean to ignore your comments, Steve :)

My previous idea to look at the `step` property only complicates things further, because you could still have a `step` specified and have a `:00` at the end of the input (if you programmatically set the `value` of the input, which is what some of the tests do). According to the [MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/time#time_value_format), the only two possible formats are indeed `HH:mm` and `HH:mm:ss` for `time` inputs. Similar idea for `datetime-local`, although MDN is a little more ambiguous about that (but I've been able to confirm similar behavior through testing).

I could leave the length check as-is, or do the regex approach if that seems better. Just to clarify, the case for `'time'` already worked using a length check (but we could change that one too).

File: src/Components/Web/src/Forms/InputDate.cs (hunk 5 above)
Comment (MackinnonBuck) [reply src/Components/Web/src/Forms/InputDate.cs:54]: Using an enum parameter like `InputDateType` was something I considered, but my primary reasoning for looking at the `type` attribute instead was mainly to keep the API surface small and to reuse existing web concepts developers might already be familiar with. However, it also makes the feature difficult to discover, and it breaks the pattern that we've been following of abstracting away the `type` attribute. So, I think having an `InputDateType` might actually be the better way to go.

File: src/Components/Web.JS/src/Rendering/Events/EventTypes.ts (hunk 2 above)
Comment (pranavkm) [reply src/Components/Web.JS/src/Rendering/Events/EventTypes.ts:266]: > Just to clarify, the case for 'time' already worked using a length check (but we could change that one too).

Oh yeah, I see that now. Well if we have precedence, leaving this as-is should be fine. I'll consider this conversation resolved.

**Review by pranavkm — APPROVED** [review]

**Comment by MarvinKlein1508** [comment]
@MackinnonBuck 
 as I mentioned in #34624 there might be an issue when you try to validate `datetime-loca`l and `time` with `CultureInfo.InvariantCulture`

Validation has worked in my case until 12:00 for time. In Europe time is defined up to 23:59

23:00 is equivalent to 11PM. 

I would suggest to check if your implementation works with 24 hour time format.

**Comment by MackinnonBuck** [comment]
@MarvinKlein1508 Could you please clarify this a little?

Date and time HTML elements (including `datetime-local` and `time`) provide a culture-invariant format via the HTML DOM API regardless of the locale of the user's browser. I have tested `<InputDate>` with a browser locale with a 24-hour time format to verify this.

What issues with validation were you experiencing and how did you produce them?

Cheers.

**Comment by MarvinKlein1508** [comment]
@MackinnonBuck 

sure. I created a small video for you which shows you what I mean. This is using the approach I mentioned in #34624

You can see it here: https://www.youtube.com/watch?v=y5DUWSMJVAA

As you can see the input gets the invalid class when entering a value above 12:59 for time. 

I just want to let you know about this. Since your solution is a different approach I thought you might have the same issue.

The component I used can be recreated with this code:
```
/// <summary>
/// An input component for editing date values.
/// Supported types are <see cref="DateTime"/> and <see cref="DateTimeOffset"/>.
/// </summary>
public class InputDate<TValue> : InputBase<TValue>
{
    /// <summary>
    /// Gets or sets the error message used when displaying an a parsing error.
    /// </summary>
    [Parameter] public string ParsingErrorMessage { get; set; } = "The {0} field must be a date.";
    
    /// <summary>
    /// Gets or sets the <see cref="InputDateFormat"/> for the input.
    /// </summary>
    [Parameter] public InputDateFormat Format { get; set; } = InputDateFormat.Date;

    /// <summary>
    /// Gets or sets the associated <see cref="ElementReference"/>.
    /// <para>
    /// May be <see langword="null"/> if accessed before the component is rendered.
    /// </para>
    /// </summary>
    [DisallowNull] public ElementReference? Element { get; protected set; }

    /// <summary>
    /// Gets the corresponding input type for the provided <see cref="Format"/>
    /// </summary>
    protected string InputType => Format switch
    {
        InputDateFormat.Date => "date",
        InputDateFormat.DateTimeLocal => "datetime-local",
        InputDateFormat.Time => "time",
        _ => throw new InvalidOperationException($"The format '{Format} is not supported'")
    };

    /// <summary>
    /// Gets the corresponding input format for the provided <see cref="Format"/>
    /// </summary>
    protected string InputFormat => Format switch
    {
        InputDateFormat.Date => "yyyy-MM-dd",
        InputDateFormat.DateTimeLocal => "yyyy-MM-ddThh:mm",
        InputDateFormat.Time => "hh:mm:ss",
        _ => throw new InvalidOperationException($"The format '{Format} is not supported'")
    };

    /// <inheritdoc />
    protected override void BuildRenderTree(RenderTreeBuilder builder)
    {
        builder.OpenElement(0, "input");
        builder.AddMultipleAttributes(1, AdditionalAttributes);
        builder.AddAttribute(2, "type", InputType);
        builder.AddAttribute(3, "class", CssClass);
        builder.AddAttribute(4, "value", BindConverter.FormatValue(CurrentValueAsString));
        builder.AddAttribute(5, "onchange", EventCallback.Factory.CreateBinder<string?>(this, __value => CurrentValueAsString = __value, CurrentValueAsString));
        builder.AddElementReferenceCapture(6, __inputReference => Element = __inputReference);
        builder.CloseElement();
    }

    /// <inheritdoc />
    protected override string FormatValueAsString(TValue? value)
    {
        switch (value)
        {
            case DateTime dateTimeValue:
                return BindConverter.FormatValue(dateTimeValue, InputFormat, CultureInfo.InvariantCulture);
            case DateTimeOffset dateTimeOffsetValue:
                return BindConverter.FormatValue(dateTimeOffsetValue, InputFormat, CultureInfo.InvariantCulture);
            default:
                return string.Empty; // Handles null for Nullable<DateTime>, etc.
        }
    }

    /// <inheritdoc />
    protected override bool TryParseValueFromString(string? value, [MaybeNullWhen(false)] out TValue result, [NotNullWhen(false)] out string? validationErrorMessage)
    {
        // Unwrap nullable types. We don't have to deal with receiving empty values for nullable
        // types here, because the underlying InputBase already covers that.
        var targetType = Nullable.GetUnderlyingType(typeof(TValue)) ?? typeof(TValue);

        bool success;
        if (targetType == typeof(DateTime))
        {
            success = TryParseDateTime(value, InputFormat, out result);
        }
        else if (targetType == typeof(DateTimeOffset))
        {
            success = TryParseDateTimeOffset(value, InputFormat, out result);
        }
        else
        {
            throw new InvalidOperationException($"The type '{targetType}' is not a supported date type.");
        }

        if (success)
        {
            Debug.Assert(result != null);
            validationErrorMessage = null;
            return true;
        }
        else
        {
            validationErrorMessage = string.Format(CultureInfo.InvariantCulture, ParsingErrorMessage, DisplayName ?? FieldIdentifier.FieldName);
            return false;
        }
    }

    private static bool TryParseDateTime(string? value, string format, [MaybeNullWhen(false)] out TValue result)
    {
        var success = BindConverter.TryConvertToDateTime(value, CultureInfo.InvariantCulture, format, out var parsedValue);
        if (success)
        {
            result = (TValue)(object)parsedValue;
            return true;
        }
        else
        {
            result = default;
            return false;
        }
    }

    private static bool TryParseDateTimeOffset(string? value, string format, [MaybeNullWhen(false)] out TValue result)
    {
        var success = BindConverter.TryConvertToDateTimeOffset(value, CultureInfo.InvariantCulture, format, out var parsedValue);
        if (success)
        {
            result = (TValue)(object)parsedValue;
            return true;
        }
        else
        {
            result = default;
            return false;
        }
    }
}
```


```
/// <summary>
/// Defines the types of an <see cref="InputDate{TValue}"/> component.
/// </summary>
public enum InputDateFormat
{
    Date,
    DateTimeLocal,
    Time
}
```

I have the same behaviour in all of my browsers. I am running Windows 10 in German with a German version of Firefox 90.

**Comment by MackinnonBuck** [comment]
Thank you for the detailed response, @MarvinKlein1508!

I have replicated your test and it appears to work fine with the implementation in this PR.

https://user-images.githubusercontent.com/10456961/127200519-f0832592-262c-41ff-8925-5ac3dc9b7f94.mp4

Thanks again for your help 😃
